test/session.test.js

```diff
@@ -0,0 +1,160 @@
+/* eslint-disable no-unused-expressions */
+/* eslint-disable no-undef */
+import * as Factory from './lib/factory.js';
+chai.use(chaiAsPromised);
+const expect = chai.expect;
+
+describe('server session', () => {
+  const BASE_ITEM_COUNT = 1; /** Default items key */
+
+  const syncOptions = {
+    checkIntegrity: true,
+    awaitAll: true
+  };
+
+  before(async function () {
+    localStorage.clear();
+  });
+
+  after(async function () {
+    localStorage.clear();
+  });
+
+  beforeEach(async function () {
+    this.timeout(Factory.TestTimeout);
+    this.expectedItemCount = BASE_ITEM_COUNT;
+    this.application = await Factory.createInitAppWithRandNamespace();
+    this.email = Uuid.GenerateUuidSynchronously();
+    this.password = Uuid.GenerateUuidSynchronously();
+    await Factory.registerUserToApplication({
+      application: this.application,
+      email: this.email,
+      password: this.password
+    });
+  });
+
+  afterEach(async function () {
+    expect(this.application.syncService.isOutOfSync()).to.equal(false);
+    const items = this.application.itemManager.items;
+    expect(items.length).to.equal(this.expectedItemCount);
+    const rawPayloads = await this.application.storageService.getAllRawPayloads();
+    expect(rawPayloads.length).to.equal(this.expectedItemCount);
+    await this.application.deinit();
+  });
+
+  function getDelayBeforeNextRequest(currentSession) {
+    const tokenExpireAt = currentSession.expireAt;
+    const timeRemaining = Math.ceil(tokenExpireAt - Date.now() / 1000);
+    /* 
+      If the token has not expired yet, we will return the remaining time, plus one second.
+      Else, there's no need to add a delay.
+    */
+    return timeRemaining > 0 ? timeRemaining + 1 : 0;
```

Comment:
I still don't think you should be adding 1s. It seems like you're adding 1s "to be safe", but we don't want to be safe, we want to be precise. What's the difference between 1ms and 1s? A session should be 100% expired 1ms after its expirey timestamp.

Reply:
Currently, the `syncing-server` is checking token expiration to the second: 

`session.rb`
```ruby
def expired_access_token?
    access_token_expire_at < DateTime.now.to_i
end
```

`DateTime.now.to_i` returns a 10-digit integer (in seconds).

So we may need to rewrite it to use a 13-digit integer (with milliseconds) in order to be more precise.

Reply:
Hmm, might be good to be more precise than. `to_f` instead of `to_i` perhaps. What are the pros/cons?

Reply:
Let's use `DateTime.now.strftime('%Q').to_i`.

It returns a timestamp with milliseconds.